ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+            (span.get_tag("langchain.request.model") or ""),
+            (span.get_tag("langchain.request.type") or ""),
+            api_key,
+        )
+        return tags
+
+    def _metrics_tags(self, span):
+        provider = span.get_tag("langchain.request.provider") or ""
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = [
+            "version:%s" % (config.version or ""),
+            "env:%s" % (config.env or ""),
+            "service:%s" % (span.service or ""),
+            "langchain.request.provider:%s" % provider,
+            "langchain.request.model:%s" % (span.get_tag("langchain.request.model") or ""),
+            "langchain.request.type:%s" % (span.get_tag("langchain.request.type") or ""),
+            "langchain.request.api_key:%s" % api_key,
+            "error:%d" % span.error,
+        ]
+        err_type = span.get_tag("error.type")
+        if err_type:
+            tags.append("error_type:%s" % err_type)
+        return tags
+
+    def record_usage(self, span, usage):
+        if not usage or not self._config.metrics_enabled:
+            return
+        for token_type in ("prompt", "completion", "total"):
+            num_tokens = usage.get("token_usage", {}).get(token_type + "_tokens")
+            if not num_tokens:
+                continue
+            self.metric(span, "dist", "tokens.%s" % token_type, num_tokens)
+        total_cost = span.get_metric("langchain.tokens.total_cost")
+        if total_cost:
+            self.metric(span, "incr", "tokens.total_cost", total_cost)
+
+
+def _extract_model_name(instance):
+    """Extract model name or ID from llm instance."""
+    for attr in ("model", "model_name", "model_id", "model_key", "repo_id"):
+        if hasattr(instance, attr):
+            return getattr(instance, attr, None)
+    return None
+
+
+def _extract_api_key(instance):
+    """
+    Extract and format LLM-provider API key from instance.
+    Note that langchain's LLM/ChatModel/Embeddings interfaces do not have a
+    standard attribute name for storing the provider-specific API key, so make a
+    best effort here by checking for attributes that end with `api_key/token`.
+    """
+    api_key_attrs = [a for a in dir(instance) if a.endswith(("api_token", "api_key"))]
+    if api_key_attrs and hasattr(instance, str(api_key_attrs[0])):
+        api_key = getattr(instance, api_key_attrs[0], None)
+        if api_key:
+            return "...%s" % api_key[-4:]
+    return ""
+
+
+def _tag_openai_token_usage(span, llm_output, propagated_cost=0, propagate=False):
+    """
+    Extract token usage from llm_output, tag on span.
+    Calculate the total cost for each LLM/chat_model, then propagate those values up the trace so that
+    the root span will store the total token_usage/cost of all of its descendants.
+    """
+    for token_type in ("prompt", "completion", "total"):
+        current_metric_value = span.get_metric("langchain.tokens.%s_tokens" % token_type) or 0
```

Comment:
I think this is fine for now since this is just 3 metrics. If we do end up needing to add more metrics I think we can consider your suggestion of using constants.